docs/_drafts/go-strategy-pattern-a-response.md

```diff
@@ -0,0 +1,172 @@
+    return message
+})
+Display(message, TextFormatted)
+```
+
+```go
+tf := &TextFormatter{}
+Display(message, tf)
+```
+
+There is another benefit to using a struct as the method receiver, which rednafi very briefly touches upon: should we need to, we can add state.
+An empty struct is very cheap to use as it requires no space to allocate, but should we need to add state then everything else is already in place[3].
+If we wanted to add state to the original approach, we would need to change it to use a struct anyway, likely causing reworking in other code that should, really, be unaffected.
+
+In summary, using **structs** to implement the strategy interface is **simpler, more concise, and more readily extensible**.
+
+## Functioning differently
+
+I mentioned that I'd propose more than one alternative to Redowan's approach.
+[Footnote 4](https://rednafi.com/go/strategy_pattern/#fn:4) to their post gave me a little brainwave --- what if there's a nicer way to handle passing around functions as strategies?
+
+If we want to bend the definition of the Strategy Pattern, we could argue that a strategy is really defined by the signature of its invocation function.
+Given that the canonical strategy interface comprises a single function, the interface itself is really just a way of assigning a type for that function in languages that don't support raw functions but rather only methods, like Java.
+
+Using this looser interpretation, we can do away with that superfluous wrapping we saw before and pass functions instead of interfaces.
+The following is a minimal working example tested with Go 1.21.5.
+
+```go
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+type Format func(message string) string
+
+type jsonFormat struct{}
+
+func (jf *jsonFormat) Format(message string) string {
+	asJSON, _ := json.Marshal(map[string]string{"message": message})
+	return string(asJSON)
+}
+
+var JSON = &jsonFormat{}
+
+func Display(message string, f Format) {
+	fmt.Println(f(message))
+}
+
+func main() {
+	message := "Hello, World!"
+	textFormat := func(message string) string { return message }
+	j := &jsonFormat{}
+
+	Display(message, textFormat)
+	Display(message, j.Format)
+	Display(message, JSON.Format)
+}
+```
+
+There are a few key differences of which to take notice.
+We now define `Format` as a function instead of using the `Formatter` interface[4].
+The `textFormat` strategy is defined as a lambda to show that this is legal syntax, even without any wrapping.
+The `jsonFormat` strategy is defined as a struct to show that we can pass in a method so long as its signature is compatible --- we have not lost the ability to use stateful strategies!
+Using a method on a struct requires instantiating it first, which might be inconvenient.
+The exported (public) `JSON` variable shows one approach to working around that inconvenience.
+
+---
+
+## Footnotes
+
+[1] I'm a big believer in writing code that expresses algorithms rather than getting lost in the weeds.
```

Comment:
:wrench: Add hyperlinks to footnotes section.  Likely also worth adding backlinks, at least to the relevant section.

Reply:
Fixed in https://github.com/agrski/agrski.github.io/pull/45/commits/5d7319a5b0fc74a0b64173e1514671a117da896f